ui/StatusQ/src/writableproxymodel.cpp

```diff
@@ -0,0 +1,878 @@
+#include "StatusQ/writableproxymodel.h"
+#include <QSet>
+#ifdef QT_DEBUG
+#include <QAbstractItemModelTester>
+#endif
+#include <memory>
+
+template <typename T>
+using IndexedValues = QHash<T /*key*/, QMap<int/*role*/, QVariant/*value*/>>;
+
+class WritableProxyModelPrivate
+{
+public:
+    explicit WritableProxyModelPrivate(WritableProxyModel& q)
+        : q(q)
+    {
+    }
+
+    WritableProxyModel& q;
+    IndexedValues<QPersistentModelIndex> cache;
+    IndexedValues<QModelIndex> insertedRows;
+    QSet<QPersistentModelIndex> removedRows;
+    QVector<int> proxyToSourceRowMapping;
+
+    inline void setData(const QModelIndex& index, const QVariant& value, int role);
+    template<typename T>
+    inline void setData(const QModelIndex& index, const QVariant& value, int role, IndexedValues<T>& indexedMap);
+
+    inline QVariant data(const QModelIndex& index, int role, bool& found) const;
+    template<typename T>
+    inline QVariant data(const QModelIndex& index, int role, bool& found, const IndexedValues<T>& indexedMap) const;
+
+    inline int proxyToSourceRow(int row) const;
+    inline int sourceToProxyRow(int row) const;
+    QVector<QPair<int, int>> sourceRowRangesBetween(int start, int end) const;
+
+    //Simple mapping. No sorting, no moving
+    //TODO: add mapping for temporarily moved rows
+    void createProxyToSourceRowMap();
+
+    inline void clear();
```

Comment:
Excluding some rare cases, the general rule of thumb is to skip `inline` keyword. If not present - compiler still can inline the function. Also when declared, compiler can ignore it. So I would opt for just skipping it (https://isocpp.org/wiki/faq/inline-functions).